submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/service/DoiService.kt

```diff
@@ -0,0 +1,125 @@
+import org.springframework.http.MediaType.MULTIPART_FORM_DATA
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.web.reactive.function.client.WebClient
+import java.nio.file.Files
+
+private val logger = KotlinLogging.logger {}
+
+@Suppress("ThrowsCount", "ReturnCount")
+class DoiService(
+    private val webClient: WebClient,
+    private val properties: DoiProperties,
+) {
+    fun calculateDoi(accNo: String, rqt: SubmitRequest): String? {
+        val doi = rqt.submission.attributes.find { it.name == DOI.name } ?: return null
+        val previousDoi = rqt.previousVersion?.doi
+
+        if (previousDoi != null) {
+            require(doi.value == previousDoi) { throw InvalidDoiException() }
+            return previousDoi
+        }
+
+        return registerDoi(accNo, rqt)
+    }
+
+    private fun registerDoi(accNo: String, rqt: SubmitRequest): String {
+        val sub = rqt.submission
+        val title = requireNotNull(sub.title) { throw MissingTitleException() }
+        val request = DoiRequest(accNo, title, properties.uiUrl, getContributors(sub))
+        val requestFile = Files.createTempFile("${TEMP_FILE_NAME}_$accNo", ".xml").toFile()
+        FileUtils.writeContent(requestFile, request.asXmlRequest())
+
+        val headers = HttpHeaders().apply { contentType = MULTIPART_FORM_DATA }
+        val body = LinkedMultiValueMap<String, Any>().apply {
+            add(USER_PARAM, properties.user)
+            add(PASSWORD_PARAM, properties.password)
+            add(OPERATION_PARAM, OPERATION_PARAM_VALUE)
+            add(FILE_PARAM, FileSystemResource(requestFile))
+        }
+
+        webClient.post(properties.endpoint, RequestParams(headers, body))
+        logger.info { "$accNo ${rqt.owner} Registered DOI: '${request.doi}'" }
+
+        return request.doi
+    }
+
+    private fun getContributors(sub: Submission): List<Contributor> {
+        val organizations = getOrganizations(sub)
+        return sub.allSections()
+            .filter { it.type.lowercase() == AUTHOR_TYPE }
+            .ifEmpty { throw MissingDoiFieldException(AUTHOR_TYPE) }
+            .map { it.asContributor(organizations) }
+    }
+
+    private fun Section.asContributor(orgs: Map<String, String>): Contributor {
+        val attrsMap = attributes.associateBy({ it.name.lowercase() }, { it.value })
+        val names = requireNotNull(attrsMap[NAME_ATTR]) { throw InvalidAuthorNameException() }
+        val affiliation = requireNotNull(attrsMap[AFFILIATION_ATTR]) { throw MissingAuthorAffiliationException() }
+        val org = requireNotNull(orgs[affiliation]) { throw InvalidAuthorAffiliationException(names, affiliation) }
+
+        return Contributor(
+            name = names.substringBeforeLast(" ", ""),
+            surname = names.substringAfterLast(" "),
+            affiliation = org,
+            orcid = attrsMap[ORCID_ATTR]
+        )
+    }
+
+    private fun getOrganizations(sub: Submission): Map<String, String> {
```

Comment:
I think this code can be simplified if we separate the two diferent extensions, (get the organizations and validate them) actually the name is confusing because it only do not get it but also validate them. 

What we can do is simple create two methods or two fractios of code, one that create the list of ornigzation and another that validate them according to requirments.